Tests: Add search mode "All"

bksahu opened the pull request:
Fixes #372 

Thank your for contributing to Nuitka!

!! Please check that you select the **develop branch** (details see below link) !!

Before submitting a PR, please review the guidelines:
https://github.com/kayhayen/Nuitka/blob/master/CONTRIBUTING.md

### What does this PR do?

### Why was it initiated? Any relevant Issues?

### PR Checklist
- [x] Correct base branch selected? `develop` for new features and bug fixes too. If it's
      part of a hotfix, it will be moved to ``master`` during it.
- [x] All tests still pass. Check the developer manual about ``Running the Tests``. There
      are Travis tests that cover the most important things however, and you are
      welcome to rely on those, but they might not cover enough.
- [ ] Ideally new features or fixed regressions ought to be covered via new tests.
- [ ] Ideally new or changed features have documentation updates.

kayhayen commented on tests/optimizations/run_all.py:
```diff
@@ -108,12 +108,15 @@ def checkSequence(statements):
-                    "%s: Error, print of non-constant '%s'." % (
-                        getSourceRef(statement),
-                        getKind(print_arg)
+                if isinstance(search_mode, SearchModeAll):
```
One style complaint here. I think using `isinstance` is nearly always too involved with classes and their hierarchy. Instead, most often, a `isXYZ` or `shallABC` method that is asked for a feature or desired behavior better.

Also, in this concrete example, I dislike the `sys.exit()` being conditional. I think that is precisely what we want the search mode to decide.

Instead call `search_mode.exit(filename, message =...)`, the default implementation will then call `sys.exit(message)` whereas the all mode can pick it up, and add it to the list of errors occurred.

kayhayen commented on nuitka/tools/testing/SearchModes.py:
```diff
@@ -173,3 +174,15 @@ def abortIfExecuted(self):
+    def updateTotalErrors(self):
+        self.total_errors += 1
+
+    def finish(self):
```
This probably should also `sys.exit()` with error code, so it's noticed by outside world.

kayhayen replied:
See however first my comment below.

kayhayen commented on nuitka/tools/testing/Common.py:
```diff
@@ -341,8 +343,11 @@ def compareWithCPython(dirname, filename, extra_flags, search_mode, needs_2to3):
     if result != 0 and result != 2 and search_mode.abortOnFinding(dirname, filename):
-        my_print("Error exit!", result)
-        sys.exit(result)
+        if search_mode.isSearchModeAll():
```
I would rather make this `search_mode.onErrorDetected(message)` and not use a test method.

When I said that test methods of the `isXY` kind are better, that was true. But even better, make the search mode decide what's going to happen entirely, and just inform it about facts.

bksahu replied:
Sorry, I don't understand. Do you want me to store and display errors instead of just incrementing the error counter?

bksahu commented:
@kayhayen I have found a weird bug. During last commit `c01d046` somehow all the files I have changed has no executable permission due to which I am getting `permission denied` error while trying to run the tests. BTW before committing, I manually ran  `autoformat-nuitka-source` and `check-nuitka-with-pylint3`.

kayhayen commented on nuitka/tools/testing/Common.py:
```diff
@@ -341,8 +343,13 @@ def compareWithCPython(dirname, filename, extra_flags, search_mode, needs_2to3):
+        # else:
+        #     message = "Error exit!" + str(result)
+        #     search_mode.exit(message)
+        message = "Error exit!" + str(result)
```
Why not use the "%" formatting?

bksahu replied:
Okay

bksahu commented:
@kayhayen So, should I close this PR?

kayhayen commented:
I was waiting for you to address my review comments, then merge it. The merge conflicts are probably from black formatting which I try to now apply to test runners as well. However, there can be an # isort:start trick missing in some,beware.

bksahu commented:
Okay. I think I have already addressed all your reviews but I will look into it again today. Also I am unable to resolve the merge conflict as it says "Only those with write access to this repository can merge pull requests."

bksahu commented:
@kayhayen I'm ready for a review.

kayhayen commented on nuitka/tools/testing/Common.py:
```diff
@@ -347,8 +349,12 @@ def compareWithCPython(dirname, filename, extra_flags, search_mode, needs_2to3):
     if result != 0 and result != 2 and search_mode.abortOnFinding(dirname, filename):
-        my_print("Error exit!", result)
-        sys.exit(result)
+        # if search_mode.isSearchModeAll():
```
You probably meant to remove the commented out codes?

kayhayen commented on nuitka/tools/testing/Common.py:
```diff
@@ -372,8 +378,11 @@ def checkCompilesNotWithCPython(dirname, filename, search_mode):
     if result != 1 and result != 2 and search_mode.abortOnFinding(dirname, filename):
-        my_print("Error exit!", result)
-        sys.exit(result)
+        if isinstance(search_mode, SearchModeAll):
```
Not using `onErrorDetected` here, but probably should too.

kayhayen commented on nuitka/tools/testing/SearchModes.py:
```diff
@@ -82,10 +82,24 @@ def _match(cls, dirname, filename, candidate):
         return False
 
+    def onErrorDetected(self, message):
+        if self.isSearchModeAll():
```
Please use overloading of methods. This should simply be the `sys.exit` I figure, while the method in the all search mode would do what it does. You do not need `isSearchModeAll` in this case, which was the point of using a method.

kayhayen commented on tests/optimizations/run_all.py:
```diff
@@ -33,12 +36,8 @@
-    )
+        os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..")
+    ),
 )
```
You are a victim here of the commit hook now covering `run_all.py`, but there are no isort:start to make sure the `nuitka` package can be imported. Please add `# isort:start` in much the same way I did in some other runners (can search for it), and it is my fault. I ought to have added these immediately to of all them.

kayhayen commented on tests/test-runners/run_all.py:
```diff
@@ -39,9 +41,11 @@
+from nuitka.tools.testing.SearchModes import SearchModeAll
 
 def main():
+    search_mode = createSearchMode()
```
This one is not using search modes at all yet, wrong to add it without a scan.

bksahu commented:
@kayhayen Should we have a testing module to test the "search modes"?

kayhayen commented:
No, I don't think that is needed. Us looking at them running will have to do. :)

kayhayen commented on .gitignore:
```diff
@@ -33,3 +33,4 @@ Changelog.html
 doc/*.1
 doc/man-*.html
 doxygen-warnings.log
+venv_cpython/
```
The lack of trailing new line (not a problem), just curious, is that something that the commit hook doesn't do? Because I think it should.

bksahu replied:
I think it doesn't

kayhayen replied:
I prepared an update on factory, which makes it handle .gitignore files too.

kayhayen commented on tests/packages/run_all.py:
```diff
@@ -19,51 +19,48 @@
+        os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..")
+    ),
+)
+
```
I noticed you did add the isort:start correctly for one runner, but basically all that to that sys.path insertion need to do it, sorry, that is a bunch.

kayhayen commented:
Looks good aside those missing `# isort:start` markers that allow the isort to handle the sys.path extension to allow finding local `nuitka` from checkout gracefully. And of course, there is also still the method `isSearchModeAll` or so, which you stopped using, but has no right to exist. 

Thanks for the good work. I am almost tempted to repair the `isort:start` myself.

bksahu commented:
Give me 10 minutes.

On Tue, 11 Jun, 2019, 10:08 PM Kay Hayen, <notifications@github.com> wrote:

> Looks good aside those missing # isort:start markers that allow the isort
> to handle the sys.path extension to allow finding local nuitka from
> checkout gracefully. And of course, there is also still the method
> isSearchModeAll or so, which you stopped using, but has no right to exist.
>
> Thanks for the good work. I am almost tempted to repair the isort:start
> myself.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Nuitka/Nuitka/pull/378?email_source=notifications&email_token=AGEXBNMQAYLK7YSWPDFDWDTPZ7IGZA5CNFSM4HKDIEE2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXNX5AY#issuecomment-500924035>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGEXBNNJHIYUNCIYB5VRWGLPZ7IGZANCNFSM4HKDIEEQ>
> .
>

kayhayen commented on tests/test-runners/run_all.py:
```diff
@@ -22,6 +22,8 @@
 import os
 import sys
 
+from nuitka.tools.testing.SearchModes import SearchModeAll
```
Here is is actually harmful, and as unused ought to have been removed. The pylint doesn't cover these files yet. I am going to fix this myself though, no big deal,as Travis doesn't run these. Maybe it ought to though, I will explore that. And also adding the test runners to pylint coverage.
